composer/checkpoint/upload.py

```diff
@@ -0,0 +1,209 @@
+
+import logging
+import os
+import pathlib
+import time
+from concurrent.futures import Future
+from typing import Optional
+
+import composer.utils.dist as dist
+from composer.core import Callback, State
+from composer.loggers import Logger, MosaicMLLogger
+from composer.utils import (
+    RemoteFilesExistingCheckStatus,
+    RemoteUploader,
+    create_symlink_file,
+    parse_uri,
+)
+
+log = logging.getLogger(__name__)
+
+
+class CheckpointUploadCallback(Callback):
+    """Callback class for async checkpoint uploading."""
+
+    def __init__(
+        self,
+        remote_uploader: RemoteUploader,
+        upload_timeout_in_seconds: int = 3600,
+    ):
+        self.remote_uploader = remote_uploader
+        if dist.get_global_rank() == 0:
+            self.rank_saves_symlinks = True
+        else:
+            self.rank_saves_symlinks = False
+
+        self.symlink_upload_tasks: list[tuple[Future[RemoteFilesExistingCheckStatus], str, Optional[str]]] = []
+        self.upload_timeout_in_seconds: int = upload_timeout_in_seconds
+
+        # Allow unit test to override this to make it faster
+        self._symlink_upload_wait_before_next_try_in_seconds = 30.0
+
+    def add_symlink_upload_task(
+        self,
+        remote_file_names: list[str],
+        symlink_file_name: Optional[str] = None,
+        symlink_remote_file_name: Optional[str] = None,
+    ):
+        check_remote_files_exist_future = None
+        if self.rank_saves_symlinks and symlink_file_name is not None:
+            check_remote_files_exist_future = self.remote_uploader.check_remote_files_exist_async(
+                remote_checkpoint_file_names=remote_file_names,
+                max_wait_time_in_seconds=self.upload_timeout_in_seconds,
+                wait_before_next_try_in_seconds=self._symlink_upload_wait_before_next_try_in_seconds,
+            )
+            self.symlink_upload_tasks.append(
+                (check_remote_files_exist_future, symlink_file_name, symlink_remote_file_name),
+            )
+
+    def _log_checkpoint_upload(self, logger: Logger):
+        for destination in logger.destinations:
+            if isinstance(destination, MosaicMLLogger):
+                destination.log_metadata({'checkpoint_uploaded_time': time.time()}, force_flush=True)
+
+    def batch_end(self, state: State, logger: Logger) -> None:
+        del state  # unused
+        self.remote_uploader.check_workers()
+        if not self.rank_saves_symlinks:
+            return
+        undone_symlink_upload_tasks = []
+        for (check_remote_files_exist_future, local_symlink_file,
+             remote_symlink_file) in reversed(self.symlink_upload_tasks):
+            if not check_remote_files_exist_future.done():
+                undone_symlink_upload_tasks.insert(
+                    0,
+                    (check_remote_files_exist_future, local_symlink_file, remote_symlink_file),
+                )
```

Comment:
wait once we upload the symlink for the last thing in the list then we don't need to upload the other ones? right?

Reply:
we still upload the checkpoint files. But for symlink file, once we upload one symlink, we don't need the previous ones, right?

Reply:
right so why do we still insert the remaining tasks into the list?

Reply:
oh, we reversely traverse the list and insert all the tasks that not finished. Once we find a finished one, we stop the traverse.